linker/shared/src/test/scala/org/scalajs/linker/SmallModulesForSplittingTest.scala

```diff
@@ -92,4 +94,87 @@ class SmallModulesForSplittingTest {
       assertEquals(5, moduleSet.modules.size)
     }
   }
+
+  @Test
+  def noCircularDepsThroughFineGrainedClasses_Issue4835(): AsyncResult = await {
+    /* Test a particular shape of dependencies that used to produce modules
+     * with cyclic dependencies. Because of that, it used to fail when creating
+     * the ModuleSet with
+     *   "requirement failed: Must have exactly one root module".
+     * With that `require` statement disabled in the `ModuleSet` constructor,
+     * it used to then fail in `checkNoCyclicDependencies`.
+     */
+
+    val SMF = EMF.withNamespace(MemberNamespace.PublicStatic)
+
+    def methodHolder(name: ClassName, methodName: String, body: Tree): ClassDef = {
+      classDef(name,
+        kind = ClassKind.Interface,
+        methods = List(
+          MethodDef(SMF, m(methodName, Nil, I), NON, Nil, IntType, Some(body))(
+              EOH.withNoinline(true), UNV)
+        ))
+    }
+
+    def call(name: ClassName, methodName: String): Tree =
+      ApplyStatic(EAF, name, m(methodName, Nil, I), Nil)(IntType)
+
+    val EntryPointsClass = ClassName("lib.EntryPoints")
+    val entryPointsClassDef = classDef(
+      EntryPointsClass,
+      superClass = Some(ObjectClass),
+      methods = List(
+        trivialCtor(EntryPointsClass),
+        mainMethodDef(consoleLog(call("lib.A", "baz")))
+      ),
+      topLevelExportDefs = List(
+        TopLevelMethodExportDef("mod",
+            JSMethodDef(SMF, str("exp"), Nil, None, call("lib.A", "baz"))(EOH, UNV))
+      )
```

Comment:
Consider just two top-level method export defs for simplicity / symmetry. FTR: The reason we need two entry points is that `lib.A` doesn't go into the public module (and `lib.B` into a different one due to public module import avoidance).

Reply:
Ah yes, that's simpler indeed. Updated.